### PR TITLE
feat:delete ollama models

### LINKED ISSUE #452: allow for deletion of ollama models
### problem

can't right now delete ollama models

### solution

- use [ollama](https://www.npmjs.com/package/ollama) npm library and call .delete method to delete model
- create ipc handler which calls this method
- forward ipc handler via preload to renderer process
- in default llm dropdown in settings have a small delete icon next to ollama/local llms that calls the delete preload function

### PR DESCRIPTION
**Pull request by Bhavyajain21**
fixes #452 
/claim #452

### CODE REVIEW

File: src/components/Settings/LLMSettings/DefaultLLMSelector.tsx [hunk 1]
```diff
@@ -25,16 +41,30 @@ const DefaultLLMSelector: React.FC<DefaultLLMSelectorProps> = ({ llmConfigs, def
     })
   }
 
+  const deleteLLM = async (modelName: string) => {
+    await window.llm.deleteLLM(modelName)
+  }
```
[the remark this answers is not in the document]
Comment (Bhavyajain21) [reply src/components/Settings/LLMSettings/DefaultLLMSelector.tsx]: Added a user feedback modal

File: src/components/Settings/LLMSettings/DefaultLLMSelector.tsx [hunk 2]
```diff
@@ -11,6 +12,21 @@ interface DefaultLLMSelectorProps {
 
 const DefaultLLMSelector: React.FC<DefaultLLMSelectorProps> = ({ llmConfigs, defaultLLM, setDefaultLLM }) => {
   const [selectedLLM, setSelectedLLM] = useState(defaultLLM)
+  const [availableModels, setAvailableModels] = useState<LLMConfig[]>([])
```
[the remark this answers is not in the document]
Comment (Bhavyajain21) [reply src/components/Settings/LLMSettings/DefaultLLMSelector.tsx]: Done!

File: src/components/Settings/LLMSettings/DefaultLLMSelector.tsx [hunk 3]
```diff
@@ -25,16 +41,30 @@ const DefaultLLMSelector: React.FC<DefaultLLMSelectorProps> = ({ llmConfigs, def
...
     <Select onValueChange={handleDefaultModelChange} value={selectedLLM}>
       <SelectTrigger className="w-[180px]">
         <SelectValue placeholder="Select default LLM" />
       </SelectTrigger>
       <SelectContent>
         {llmConfigs.map((config) => (
-          <SelectItem key={config.modelName} value={config.modelName}>
-            {config.modelName}
-          </SelectItem>
+          <div className="flex w-full items-center justify-end">
```
[the remark this answers is not in the document]
Comment (Bhavyajain21) [reply src/components/Settings/LLMSettings/DefaultLLMSelector.tsx]: Done!

File: electron/main/llm/ipcHandlers.ts [hunk 4]
```diff
@@ -38,4 +38,33 @@ export const registerLLMSessionHandlers = (store: Store<StoreSchema>) => {
...
+
+      await ollamaService.deleteModel(modelName)
+      // eslint-disable-next-line @typescript-eslint/naming-convention
+      const { [modelName]: _, ...updatedModels } = currentModels
+      store.set(StoreKeys.EmbeddingModels, updatedModels)
+
+      return { success: true }
+    } catch (error: any) {
+      console.error(`Failed to delete model: ${modelName}`, error)
+      return { success: false, error: error.message }
```
[the remark this answers is not in the document]
Comment (Bhavyajain21) [reply electron/main/llm/ipcHandlers.ts:67]: This is handled

File: electron/main/llm/ipcHandlers.ts [hunk 5]
```diff
@@ -38,4 +38,33 @@ export const registerLLMSessionHandlers = (store: Store<StoreSchema>) => {
...
+      return models
+    } catch (error) {
+      console.error('Error fetching available models:', error)
+      throw error
+    }
+  })
+
+  ipcMain.handle('delete-llm', async (event, modelName: string) => {
+    try {
+      const currentModels = store.get(StoreKeys.EmbeddingModels) || {}
```
[the remark this answers is not in the document]
Comment (Bhavyajain21) [reply electron/main/llm/ipcHandlers.ts]: Resolved!

File: src/components/Settings/LLMSettings/DefaultLLMSelector.tsx [hunk 6]
```diff
@@ -11,6 +12,20 @@ interface DefaultLLMSelectorProps {
 
 const DefaultLLMSelector: React.FC<DefaultLLMSelectorProps> = ({ llmConfigs, defaultLLM, setDefaultLLM }) => {
   const [selectedLLM, setSelectedLLM] = useState(defaultLLM)
+  const [availableOllamaModels, setAvailableOllamaModels] = useState<LLMConfig[]>([])
+
+  useEffect(() => {
+    const fetchModels = async () => {
+      try {
+        const models = await window.llm.getAvailableModels()
```
Comment (joseplayero) [inline src/components/Settings/LLMSettings/DefaultLLMSelector.tsx:20]: why do we need this call at all?

imo we should use the ```llmConfigs``` prop here and filter for the llms which have the ollama api interface: 
```  apiInterface: APIInterface```

File: electron/main/llm/ipcHandlers.ts [hunk 7]
```diff
@@ -38,4 +38,32 @@ export const registerLLMSessionHandlers = (store: Store<StoreSchema>) => {
     }
     await ollamaService.pullModel(modelName, handleProgress)
   })
+
+  ipcMain.handle('get-available-models', async () => {
```
Comment (joseplayero) [inline electron/main/llm/ipcHandlers.ts:42]: i dont think we need this ipc handler at all

File: electron/main/llm/ipcHandlers.ts [hunk 8]
```diff
@@ -38,4 +38,32 @@ export const registerLLMSessionHandlers = (store: Store<StoreSchema>) => {
...
+      return models
+    } catch (error) {
+      console.error('Error fetching available models:', error)
+      throw error
+    }
+  })
+
+  ipcMain.handle('delete-llm', async (event, modelName: string) => {
+    try {
+      const currentModels: LLMConfig[] = store.get(StoreKeys.LLMs) || []
```
Comment (joseplayero) [inline electron/main/llm/ipcHandlers.ts:54]: no need to complicate this ipc handler at all

it should literally only call the delete function from ollama and not check whether the model exists in the store or not

File: electron/main/llm/ipcHandlers.ts [hunk 9]
```diff
@@ -38,4 +38,32 @@ export const registerLLMSessionHandlers = (store: Store<StoreSchema>) => {
...
+      const foundModel = currentModels.find((model) => model.modelName === modelName)
+      if (!foundModel) {
+        throw new Error(`Model ${modelName} not found in the store`)
+      }
+      const updatedModels = currentModels.filter((model) => model.modelName !== modelName)
+      store.set(StoreKeys.LLMs, updatedModels)
+      await ollamaService.deleteModel(modelName)
+
+      return { success: true }
+    } catch (error: any) {
```
Comment (joseplayero) [inline electron/main/llm/ipcHandlers.ts:64]: in reor there is no need for error handling like this. all errors are caught at the top level so you should never catch an error and log it unless an error is expected behaviour

File: electron/main/llm/ipcHandlers.ts [hunk 10]
```diff
@@ -38,4 +38,32 @@ export const registerLLMSessionHandlers = (store: Store<StoreSchema>) => {
...
+      console.error(`Failed to delete model: ${modelName}`, error)
+      return { success: false, error: error.message }
```
Comment (joseplayero) [inline electron/main/llm/ipcHandlers.ts:66]: the ipc handler does not need to return this

File: src/components/Settings/LLMSettings/DefaultLLMSelector.tsx [hunk 11]
```diff
@@ -25,16 +40,40 @@ const DefaultLLMSelector: React.FC<DefaultLLMSelectorProps> = ({ llmConfigs, def
...
+  const deleteLLM = async (modelName: string) => {
+    const confirmDelete = window.confirm(`Are you sure you want to delete the model ${modelName}?`)
+    if (!confirmDelete) return
+    try {
+      await window.llm.deleteLLM(modelName)
+      posthog.capture('delete_llm', {
+        modelName,
+      })
+    } catch (error) {
+      console.error(`Failed to delete model: ${modelName}`, error)
```
Comment (joseplayero) [inline src/components/Settings/LLMSettings/DefaultLLMSelector.tsx:52]: there should be no error catching here. just make the call outside try catch block

**Review by joseplayero — CHANGES_REQUESTED** [review]

File: src/components/Settings/LLMSettings/DefaultLLMSelector.tsx (hunk 11 above)
Comment (Bhavyajain21) [reply src/components/Settings/LLMSettings/DefaultLLMSelector.tsx]: FIxed!

File: electron/main/llm/ipcHandlers.ts (hunk 10 above)
Comment (Bhavyajain21) [reply electron/main/llm/ipcHandlers.ts]: resolved

File: electron/main/llm/ipcHandlers.ts (hunk 9 above)
Comment (Bhavyajain21) [reply electron/main/llm/ipcHandlers.ts]: Got it, resolved!

File: electron/main/llm/ipcHandlers.ts (hunk 8 above)
Comment (Bhavyajain21) [reply electron/main/llm/ipcHandlers.ts]: Correct, resolved

File: electron/main/llm/ipcHandlers.ts (hunk 7 above)
Comment (Bhavyajain21) [reply electron/main/llm/ipcHandlers.ts]: resolved!

File: src/components/Settings/LLMSettings/DefaultLLMSelector.tsx (hunk 6 above)
Comment (Bhavyajain21) [reply src/components/Settings/LLMSettings/DefaultLLMSelector.tsx]: Resolved!

**Comment by Bhavyajain21** [comment]
@joseplayero , I've worked on the addressed review comments. Please take a look!

**Review by joseplayero — APPROVED** [review]

**Pull request merged**